### DISCUSSION TITLE
Publishing reusable field types

### DISCUSSION THREAD

**Discussion by IlyaSemenov** [romeerez/orchid-orm#296] [General]
I noticed that I ended up copy/pasting `createBaseTable` options between projects, which is a pain to support and synchronize. I want to publish a small set of reusable types to npm. For instance:

```ts
uuidv7: () => t.uuid().default(uuidv7).primaryKey(),
now: () => t.timestamp().default(t.sql`now()`),
// base58 public key, might add validation later
publicKey: () => t.varchar(64),
decimal: () => t.decimal().encode((value: number | string | Decimal) => `${value}`),
text: () => t.text(0, Number.POSITIVE_INFINITY), // the infamous one
/** Store empty strings as database NULL, but never return JS null. */
optionalString: (limit?: number) => t.varchar(limit).nullable().parse((value) => value || "").encode((value: string) => value || null),
```

Question is, how would you suggest to organize that?

At first I was thinking:

```ts
import * as fields from "my-orchid-fields"

export const BaseTable = createBaseTable({
  columnTypes: (t) => ({
    ...t, // base types
    id: fields.uuidv7(t), // new type from library. odd syntax, but allows tree shaking.
  }),
})
```

can it be made simpler? how do I type my functions?

maybe this?

```ts
import { extendTypes } from "my-orchid-fields"

export const BaseTable = createBaseTable({
  columnTypes: (t) => extendTypes(t, (t2) => ({
    ...t, // base types
    id: t2.uuidv7(), // new type from library
  })),
})
```

---

This whole approach with the callbacks is kinda awkward. Ideally, I'd imagine this as composable/tree-shakeable standalone functions:

```ts
import { columnTypes } from "orchid-orm"
import { columnTypes as myColumnTypes } from "my-orchid-fields"

export const BaseTable = createBaseTable({
  columnTypes: {
    // Copy built-in field as is
    boolean: columnTypes.boolean,
    // Bulk copy
    ...pick(["numeric", "text"], columnTypes),
    // Or just copy all
    ...columnTypes,
    // Custom field based on built-in field
    varchar: (limit = 100) => columnTypes.varchar(limit),
    // Copy external field
    publicKey: myColumnTypes.publicKey,
    // Rename external field
    id: myColumnTypes.uuidv7,
    // Custom field based on external field
    nullableNow: () => myColumnTypes.now().nullable(),
  }
})
```

**Comment by romeerez** [comment]
IMO it's < 10 LOC, and even if it was 20 LOCs - not worth extracting into a package and further managing it, so I'd rely on copy-pasting as on a simple and straightforward way.

OrchidORM is designed in a way that your validation schema can be configured separately, and column types separately.

So you can use:

```ts
export const BaseTable = createBaseTable({
  schemaConfig: zodSchemaConfig, // valibotSchemaConfig, or other schema configs, may be added in future
  columnTypes: (t) => ({
    // ...column overrides
  })
})
```

Currently, validation schema has a serious downside that they can't be imported to frontend, because it's tightly coupled with ORM, but eventually I'll find a way and they will become much more useful.

So when I design something related to column types, I design it so the new column type may be slightly different between different validation schemas. Such as `json` type: without validation schema it has simple signature, and with validation schemas it can accept a `z.object`.

You may not need these validation schemas at all and your column types can be extracted in a way to support no validation schema case only.

As for three-shaking, a few extra LOCs won't harm, but I see the `Decimal` and `uuidv7` likely to use separate packages, so need to be careful that if it's not used, it's not included into bundle.

Your first example looks good to me, simple and tree-shakable, so I don't see a need for `extendTypes` or other additional things.
